### PR TITLE
Add XIF memory interface

### PR DESCRIPTION
**Pull request by michael-platzer**
This PR adds the XIF memory interfaces (i.e., the [memory request/response](https://docs.openhwgroup.org/projects/openhw-group-core-v-xif/x_ext.html#memory-request-response-interface) and the [memory result](https://docs.openhwgroup.org/projects/openhw-group-core-v-xif/x_ext.html#memory-result-interface) interfaces) to CV32E40X.

The first commit replaces the [`size` field in the XIF interface definitions](https://github.com/openhwgroup/cv32e40x/blob/380b9fd6dff586d332886e0eaa875e9aafa54ca6/rtl/if_xif.sv#L86) with a byte enable mask as discussed here: https://github.com/openhwgroup/core-v-xif/issues/17

The second commit implements the base functionality required to handle memory requests by a coprocessor. Memory requests received via the XIF memory interface are converted into memory transactions. The aligner performs alignment for XIF memory data based on the lower bits of the address as usual, with the assumption that an XIF request always uses the entire width of the memory interface (i.e., is always 32 bits wide). Note that the byte enable mask received via the XIF interface is aligned as well. Split accesses are performed if required (as is done for internal memory transactions).

The third commit forwards MPU errors and bus errors resulting from XIF memory transactions to the coprocessor. MPU errors must be reported immediately via the XIF memory response interface, rather than being buffered as is done for internal transactions. This required some changes to the MPU, which has now the option to issue MPU errors via the new `core_mpu_err_o` port immediately.

As with previous patches related to the XIF interface, the new functionality is only enabled when `X_EXT` is 1. In particular, if `X_EXT` is 0, then the functionality of CV32E40X is not modified.

Acknowledgements: The changes in this PR have been inspired by earlier work by Kunal Dalal, Oana Lazar, Yu Xia, and Adomas Lebedys from the University of Southampton in the context of the AVA vector accelerator project, see: https://github.com/AI-Vector-Accelerator

### CODE REVIEW

**Comment by Silabs-ArjanB** [comment]
Hello @michael-platzer , thank you again for your great contributions! We will review it and then merge it. In the mean time could you please have a look at the failing Eclipse ECA check; did ou maybe not sign the commit or use a different github account?

**Comment by Silabs-ArjanB** [comment]
@michael-platzer Thanks!

File: rtl/cv32e40x_load_store_unit.sv
Comment (silabs-robin) [inline rtl/cv32e40x_load_store_unit.sv:155]: Is this conditional always used, regardless of X_EXT param?
Should it be a generate block, or have an assert that `!X_EXT |-> !xif_req`?

Comment (michael-platzer) [reply rtl/cv32e40x_load_store_unit.sv:155]: Yes, this conditional is always used, but the assignment of `xif_req` is guarded by `X_EXT`:

https://github.com/openhwgroup/cv32e40x/blob/3aee97ac2e58b226e4184c57318ac4ef72900cc4/rtl/cv32e40x_load_store_unit.sv#L150

I you prefer, I can add the `X_EXT` parameter to all conditionals involving `xif_req` instead, although that might clutter the code. An assertion might be a good idea though, as well as for `xif_res_q`.

Comment (michael-platzer) [reply rtl/cv32e40x_load_store_unit.sv:155]: @silabs-robin I have tentatively added assertions for the `xif_req` and `xif_res_q` signals in the corresponding `*_sva.sv` file. Please let me know if this is correct and whether I have overlooked something, since I have no experience with verification.

File: rtl/cv32e40x_load_store_unit.sv
Comment (Silabs-ArjanB) [inline rtl/cv32e40x_load_store_unit.sv:150]: & -> && (we use && for logical and)

File: rtl/cv32e40x_load_store_unit.sv
Comment (Silabs-ArjanB) [inline rtl/cv32e40x_load_store_unit.sv:283]: Can you use a generate block here based on X_EXT and set the _q values to 0 in the else part?

File: rtl/cv32e40x_load_store_unit.sv
Comment (Silabs-ArjanB) [inline rtl/cv32e40x_load_store_unit.sv:499]: ~xif_res_q -> !xif_res_q (we use ! for single bit inversions)

File: rtl/cv32e40x_load_store_unit.sv
Comment (Silabs-ArjanB) [inline rtl/cv32e40x_load_store_unit.sv:529]: Can you simply add a 'todo' here as this seems to be the only place where valid_o_i could not be replaced by valid_0_i || xif_req. 

Can valid_0_i and xif_req both be 1 at the same time in the current design? If not, then it would be better to remove the && !xif_req part and add an assertion instead that both bits cannot be 1 at the same time. If they can be 1 at the same time can you please add a comment on for what scenario that could happen and why xif should take precedence?

File: rtl/cv32e40x_load_store_unit.sv
Comment (Silabs-ArjanB) [inline rtl/cv32e40x_load_store_unit.sv:660]: ~xif_req -> !xif_req

File: rtl/cv32e40x_load_store_unit.sv
Comment (Silabs-ArjanB) [inline rtl/cv32e40x_load_store_unit.sv:748]: Can you use a generate statement here based on X_EXT?

**Review by Silabs-ArjanB — CHANGES_REQUESTED** [review]
Hi @michael-platzer Thanks for this PR and apologies for the slow feedback. I am just requesting some minor changes here. I will also ask Øivind to double check the MPU related changes (but those look good to me). After these minor changes I will merge this. At the same time I will send you an email with respect to an alternative proposal related to the 'be'/'size' discussion (we can make change to the RTL later if you agree with that proposal).

File: rtl/cv32e40x_load_store_unit.sv
Comment (michael-platzer) [reply rtl/cv32e40x_load_store_unit.sv:536]: Yes, `valid_0_i` and `xif_req` can both be 1 at the same time when an XIF load/store instruction is immediately followed by a regular load/store instruction: In that situation, simultaneous assertion of `valid_0_i` and `xif_req` happens when the XIF load/store awaits a result in the WB stage, the regular load/store has moved to the EX stage, and the coprocessor starts the memory transaction. Loads/stores should happen in program order, hence the XIF transaction needs to take precedence.

Note that simultaneous assertion of `valid_0_i` and `xif_req` should not happen in the reverse situation (regular load/store instruction immediately followed by an XIF load/store instruction), since CV32E40X needs to wait for the memory response of the regular load/store before committing the XIF load/store (and the coprocessor must not start the XIF memory request until the instruction has been committed).

I can either add a comment explaining that situation or a TODO, please let me know which one you prefer. Do you have some prospective task in mind that I should add to the TODO (such as merging the `valid_0_i` and `xif_req` signals)?

**Comment by michael-platzer** [comment]
Thanks @Silabs-ArjanB for your feedback! I have applied the requested changes related to the logical operators and the generate blocks. Please see my comment above regarding the simultaneous assertion of `valid_0_i` and `xif_req`.

I am looking forward to hearing about your alternative proposal w.r.t the `size` vs `be` field in the XIF memory request.

File: rtl/cv32e40x_load_store_unit.sv
Comment (Silabs-ArjanB) [reply rtl/cv32e40x_load_store_unit.sv:536]: Hi @michael-platzer, thanks. I will just add above info into a new github ticket related to the introduction of a XIF related scoreboard. I expect that eventually we will make sure that valid-0_i and xif_req cannot be 1 at the same time. For now it is good enough to understand that it can actually happen.

**Review by Silabs-ArjanB — APPROVED** [review]

**Comment by silabs-oivind** [comment]
> Hi @michael-platzer Thanks for this PR and apologies for the slow feedback. I am just requesting some minor changes here. I will also ask Øivind to double check the MPU related changes (but those look good to me). After these minor changes I will merge this. At the same time I will send you an email with respect to an alternative proposal related to the 'be'/'size' discussion (we can make change to the RTL later if you agree with that proposal).

Hi @michael-platzer. The MPU changes look good to me. But the MPU is instantiated in both the LSU and the IF stage. You need to update the MPU interface for the new signals in cv32e40x_if_stage as well.

**Comment by michael-platzer** [comment]
@silabs-oivind Apologies for overlooking the second MPU in the IF stage and thanks @Silabs-ArjanB for fixing this!

**Pull request merged**
